### PR TITLE
ARROW-2627: [Python] Add option to pass memory_map argument to ParquetDataset

### PR DESCRIPTION
**Pull request by wesm**
I'm hesitant to spend much more energy on this code before diving into a C++ porting effort around this

### CODE REVIEW

File: python/pyarrow/parquet.py [hunk 1]
```diff
@@ -432,14 +432,18 @@ class ParquetDatasetPiece(object):
...
+    open_file_func : callable
+        Function to use for obtaining file handle to dataset piece
     partition_keys : list of tuples
       [(column name, ordinal index)]
     row_group : int, default None
         Row group to load. By default, reads all row groups
     """
 
-    def __init__(self, path, row_group=None, partition_keys=None):
+    def __init__(self, path, open_file_func=lambda p: open(p, 'rb'),
```
Comment (xhochy) [inline python/pyarrow/parquet.py]: Don't store `lambda` functions in attributes but rather use `partial(open, mode='rb')`. If this object is ever pickled, this makes a huge performance difference.

Comment (wesm) [reply python/pyarrow/parquet.py]: noted, thank you

**Comment by wesm** [comment]
will merge once the build runs

**Comment by wesm** [comment]
+1

**Pull request closed without merging**